[RISCV] Handle zeroinitializer of vector tuple Type

4vtomat opened the pull request:
It doesn't make sense to add a new generic ISD to handle riscv tuple
type. Instead we use `SPLAT_VECTOR` for ISD and further lower to `VMV_V_X`.

Note: If there's `visitSPLAT_VECTOR` in generic DAG combiner, it needs
to skip riscv vector tuple type.

Stack on https://github.com/llvm/llvm-project/pull/114329

4vtomat committed:
[RISCV] Handle zeroinitializer of vector tuple Type

It doesn't make sense to add a new generic ISD to handle riscv tuple
type. Instead we use `SPLAT_VECTOR` for ISD and further lower to `VMV_V_X`.

Note: If there's `visitSPLAT_VECTOR` in generic DAG combiner, it needs
to skip riscv vector tuple type.

wangpc-pp reviewed (commented):
Then we should add `TargetExtType::HasZeroInit` attribute to `riscv.vector.tuple` type in `llvm/lib/IR/Type.cpp:getTargetTypeInfo`?

4vtomat commented:
> Then we should add `TargetExtType::HasZeroInit` attribute to `riscv.vector.tuple` type in `llvm/lib/IR/Type.cpp:getTargetTypeInfo`?

Yeah, I think I added? Or do you mean we shouldn't add?

wangpc-pp commented:
> > Then we should add `TargetExtType::HasZeroInit` attribute to `riscv.vector.tuple` type in `llvm/lib/IR/Type.cpp:getTargetTypeInfo`?
> 
> Yeah, I think I added? Or do you mean we shouldn't add?

Oh, never mind! Somehow the changes in `Type.cpp` escaped my eyes. :-)

wangpc-pp commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp:
```diff
@@ -6288,6 +6288,9 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
     break;
   case ISD::SPLAT_VECTOR:
+    // RISC-V vector tuple type is not a vector type.
+    if (VT.isRISCVVectorTuple())
```
I have no opinion of this, but this may be too target-specific.

4vtomat replied:
Hmm, but I guess there's no way to skip this check lol~

lukel97 commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -1900,6 +1900,13 @@ SDValue SelectionDAGBuilder::getValueImpl(const Value *V) {
+      assert(C->isNullValue() && "Can only zero this target type!");
+      return NodeMap[V] = DAG.getNode(
+                 ISD::SPLAT_VECTOR, getCurSDLoc(), VT,
+                 DAG.getConstant(0, getCurSDLoc(), MVT::getIntegerVT(8)));
```
I'm not familiar with the RVV tuple types, but is it possible to create a SPLAT_VECTOR with a wide vector type and then perform a bitcast if that's easier?

4vtomat replied:
RVV tuple type encode `NF` information, so if we just create a normal vector type `SPLAT_VECTOR`, we are missing the information and doesn't know how to perform bitcast?

topperc replied:
You can create a normal SPLAT_VECTOR and bitcast it to the tuple type. In RISCVISelDAGToDAG:select, you can detect the (tupletype (bitcast (splat_vector 0))) and emit the correct number of instructions.

4vtomat replied:
Bitcast need both side of type size are the same, is it possible to cast a scalable vector type to a non-scalable type?

topperc replied:
Isn't the tuple type the same size as some scalable vector type?

4vtomat replied:
But its `TypeSize` is scalable, even it's same size, it's not the same.

topperc replied:
> But its `TypeSize` is scalable, even it's same size, it's not the same.

The tuple TypeSize should also be scalable since its size depends on vscale.

preames commented:
I'm clearly missing some context here, but... why do we need anything specific here to RISCVTupleType at all?  At the IR level, why is this not simply a struct?  I'd have assumed we already had the lowering for a first class aggregate, what am I missing here?

topperc commented:
> I'm clearly missing some context here, but... why do we need anything specific here to RISCVTupleType at all? At the IR level, why is this not simply a struct? I'd have assumed we already had the lowering for a first class aggregate, what am I missing here?

It was a struct a couple months ago. Unfortunately, SelectionDAGBuilder decomposes structs into individual fields before call/return handling and before inline assembly handling. This made it very difficult to have an ABI for tuples or handle tuple operands in inline assembly.

See #97992 and #97993 and #97994 and #97995

preames commented:
> > I'm clearly missing some context here, but... why do we need anything specific here to RISCVTupleType at all? At the IR level, why is this not simply a struct? I'd have assumed we already had the lowering for a first class aggregate, what am I missing here?
> 
> It was a struct a couple months ago. Unfortunately, SelectionDAGBuilder decomposes structs into individual fields before call/return handling and before inline assembly handling. This made it very difficult to have an ABI for tuples or handle tuple operands in inline assembly.
> 
> See #97992 and #97993 and #97994 and #97995

That was the context I was missing, thanks.

4vtomat commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -66,6 +66,25 @@ void RISCVDAGToDAGISel::PreprocessISelDAG() {
+      if (VT.isRISCVVectorTuple()) {
+        unsigned NF = VT.getRISCVVectorTupleNumFields();
+        unsigned NumScalElts = VT.getSizeInBits() / (NF * 8);
+        SDValue EltVal = CurDAG->getConstant(0, DL, Subtarget->getXLenVT());
```
Yeah, good idea!

4vtomat committed:
fixup! [RISCV] Handle zeroinitializer of vector tuple Type

4vtomat commented:
I added a `ISD::BITCAST` case in `PreprocessISelDAG`, but somehow the git diff made it looks like I also changed the code in `ISD::SPLAT_VECTOR` case lol

topperc commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -65,6 +87,7 @@ void RISCVDAGToDAGISel::PreprocessISelDAG() {
           VT.isInteger() ? RISCVISD::VMV_V_X_VL : RISCVISD::VFMV_V_F_VL;
       SDLoc DL(N);
       SDValue VL = CurDAG->getRegister(RISCV::X0, Subtarget->getXLenVT());
+
```
Did you insert a blank line here?

4vtomat replied:
Oh, by accident lol

lukel97 commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -57,6 +57,28 @@ void RISCVDAGToDAGISel::PreprocessISelDAG() {
 
     SDValue Result;
     switch (N->getOpcode()) {
+    case ISD::BITCAST: {
```
Do we still need the preprocessing for the bitcast? What happens if this bit is removed

4vtomat replied:
If this is removed, we need to process this in `select` function which need to expand `VMV_V_X_VL` manually which is more complicated

topperc replied:
There is a subregister hiearchy for tuples. But if the NF isn't a valid LMUL, I don't know if it would work.

4vtomat replied:
But I think we don't model the hierarchy like `VRN2M2 <-> VRM4` right?

topperc replied:
> But I think we don't model the hierarchy like VRN2M2 <-> VRM4 right?

`copyPhysReg` might work if it just looks up the start and number of registers for each of the register classes. But there's no equivalence when NF is not a power of 2.

4vtomat replied:
I think it can't, it needs both src and dst registers to be in same register class

4vtomat commented on llvm/lib/CodeGen/SelectionDAG/SelectionDAGBuilder.cpp:
```diff
@@ -1896,6 +1896,17 @@ SDValue SelectionDAGBuilder::getValueImpl(const Value *V) {
+                 DAG.getNode(
+                     ISD::SPLAT_VECTOR, getCurSDLoc(),
+                     MVT::getScalableVectorVT(
+                         MVT::i8, VT.getSizeInBits().getKnownMinValue() / 8),
```
Yeah, it will create an invalid MVT, so looks unless we have non power of 2 scalable MVT, we can't use this approach?

topperc replied:
You can create the bitcast+splat and convert it to the RISCVISD nodes in an early DAGCombine?

4vtomat replied:
But what's the return type of the splat?

topperc replied:
You can use an illegal type. There's a DAG Combiner run before the first type legalization.

4vtomat replied:
But the splat doesn't accept `INVALID_SIMPLE_VALUE_TYPE` it would crash lol

topperc replied:
You should use EVT::getScalableVectorVT not MVT::getScalableVectorVT

4vtomat replied:
I see. After combined to RISCVISD, we should still expand to NF tuple insertions manually right?

topperc replied:
> I see. After combined to RISCVISD::TUPLE_INSERT, we should still expand to NF tuple insertions manually right?

yes

4vtomat committed:
fixup! [RISCV] Handle zeroinitializer of vector tuple Type

4vtomat committed:
fixup! [RISCV] Handle zeroinitializer of vector tuple Type

4vtomat commented on llvm/lib/Target/RISCV/RISCVISelLowering.cpp:
```diff
@@ -18008,6 +18008,22 @@ SDValue RISCVTargetLowering::PerformDAGCombine(SDNode *N,
+      SDValue EltVal = DAG.getConstant(0, DL, Subtarget.getXLenVT());
+      MVT ScalTy = MVT::getScalableVectorVT(MVT::getIntegerVT(8), NumScalElts);
+
+      SDValue Splat = DAG.getNode(RISCVISD::VMV_V_X_VL, DL, ScalTy,
```
Yes!

4vtomat committed:
fixup! [RISCV] Handle zeroinitializer of vector tuple Type

topperc reviewed (approved):
LGTM
